web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.test.tsx

```diff
@@ -0,0 +1,223 @@
+      clusterPreferences: {},
+      unifiedResourcePreferences: {
+        viewMode: ViewMode.VIEW_MODE_LIST,
+        defaultTab: DefaultTab.DEFAULT_TAB_PINNED,
+      },
+    };
+
+    await act(() => result.current.updateUserPreferences(newPreferences));
+
+    // updating state
+    expect(
+      appContext.workspacesService.setUnifiedResourcePreferences
+    ).toHaveBeenCalledWith(
+      cluster.uri,
+      newPreferences.unifiedResourcePreferences
+    );
+    expect(result.current.userPreferences.unifiedResourcePreferences).toEqual(
+      newPreferences.unifiedResourcePreferences
+    );
+
+    expect(result.current.userPreferencesAttempt.status).toBe('success');
+    expect(appContext.tshd.updateUserPreferences).toHaveBeenCalledWith({
+      clusterUri: cluster.uri,
+      userPreferences: newPreferences,
+    });
+  });
+
+  test('when the request for fetching preferences is in-flight', async () => {
+    let resolveGetUserPreferencesPromise: (u: UserPreferences) => void;
+    const getUserPreferencesPromise = new Promise<UserPreferences>(resolve => {
+      resolveGetUserPreferencesPromise = resolve;
+    });
+    let resolveUpdateUserPreferencesPromise: (u: UserPreferences) => void;
+    const updateUserPreferencesPromise = new Promise(resolve => {
+      resolveUpdateUserPreferencesPromise = resolve;
+    });
+
+    jest
+      .spyOn(appContext.tshd, 'getUserPreferences')
+      .mockImplementation(() => getUserPreferencesPromise);
+    jest
+      .spyOn(appContext.tshd, 'updateUserPreferences')
+      .mockImplementation(() => updateUserPreferencesPromise);
+
+    const { result } = renderHook(() => useUserPreferences(cluster.uri), {
+      wrapper: ({ children }) => (
+        <MockAppContextProvider appContext={appContext}>
+          {children}
+        </MockAppContextProvider>
+      ),
+    });
+
+    const newPreferences: UserPreferences = {
+      clusterPreferences: {},
+      unifiedResourcePreferences: {
+        viewMode: ViewMode.VIEW_MODE_LIST,
+        defaultTab: DefaultTab.DEFAULT_TAB_PINNED,
+      },
+    };
+
+    act(() => {
+      result.current.updateUserPreferences(newPreferences);
+    });
+
+    // updating state
+    expect(
+      appContext.workspacesService.setUnifiedResourcePreferences
+    ).toHaveBeenCalledWith(
+      cluster.uri,
+      newPreferences.unifiedResourcePreferences
+    );
+    expect(result.current.userPreferences.unifiedResourcePreferences).toEqual(
+      newPreferences.unifiedResourcePreferences
+    );
+
+    expect(result.current.userPreferencesAttempt.status).toBe('processing');
+    expect(appContext.tshd.updateUserPreferences).toHaveBeenCalledWith({
+      clusterUri: cluster.uri,
+      userPreferences: newPreferences,
+    });
+
+    act(() => resolveGetUserPreferencesPromise(null));
+    await act(() => getUserPreferencesPromise);
+    act(() => resolveUpdateUserPreferencesPromise(null));
+    await act(() => updateUserPreferencesPromise);
```

Comment:
Since the initial fetch request is aborted, I couldn't test exactly that (trying to resolve a rejected promise was throwing an error). 
I added checks that verify if the users's unified resource preferences are not overwritten when a different preferences are returned from `update`.